app/models/user_track.rb

```diff
@@ -29,18 +29,47 @@ def self.for!(user_param, track_param)
     )
   end
 
-  def self.for(user_param, track_param, external_if_missing: false)
+  def self.for(user_param, track_param)
     for!(user_param, track_param)
   rescue ActiveRecord::RecordNotFound
-    return nil unless external_if_missing
-
     begin
       External.new(Track.for!(track_param))
     rescue ActiveRecord::RecordNotFound
       nil
     end
   end
 
+  memoize
+  def exercises
+    filter_enabled_exercises(track.exercises)
+  end
+
+  memoize
+  def concept_exercises
+    filter_enabled_exercises(track.concept_exercises)
+  end
+
+  memoize
+  def practice_exercises
+    filter_enabled_exercises(track.practice_exercises)
+  end
+
+  def concept_exercises_for(concept: nil)
+    filter_enabled_exercises(concept.concept_exercises) if concept.present?
+  end
+
+  def practice_exercises_for(concept: nil)
+    filter_enabled_exercises(concept.practice_exercises) if concept.present?
+  end
+
+  def unlocked_concepts_for(exercise: nil)
+    exercise.unlocked_concepts.to_a.filter { |c| concept_unlocked?(c) } if exercise.present?
+  end
+
+  def unlocked_exercises_for(exercise: nil)
+    exercise.unlocked_exercises.to_a.filter { |e| exercise_unlocked?(e) } if exercise.present?
+  end
```

Comment:
For all of these it seems like the concept/exercise is required. So should we remove the `nil` in the method definition and remove the `if` statements.